pkg/platform/kube/functionres/lazy.go

```diff
@@ -570,6 +607,106 @@ func (lc *lazyClient) waitFunctionDeploymentReadiness(ctx context.Context,
+// returns (IsDone, reasonNotDone, error)
+// Each pair explanation:
+// (true, "", nil) - all init containers are terminated with 0 exit code, we can proceed to other checks
+// (false, notReadyReason, nil) - some init containers are still waiting/running OR required number of pods hasn't been created yet, so need to wait
+// (false, "", err) - we can stop waiting here since something is broken, so function won't be successfully started
+// (true, err) - impossible
+func (lc *lazyClient) checkFunctionInitContainersDone(ctx context.Context, function *nuclioio.NuclioFunction) (bool, string, error) {
+	functionDeployment, err := lc.getFunctionDeployment(ctx, function)
+	notReadyReason := ""
+	if err != nil {
+		return false, "", err
+	}
+	// deployment doesn't exist yet
+	if functionDeployment == nil {
+		notReadyReason = "deployment doesn't exist yet"
+		return false, notReadyReason, nil
+	}
+
+	// if initContainers aren't defined or replicas number is equal to 0, skip
+	if len(function.Spec.InitContainers) == 0 || functionDeployment.Spec.Replicas == nil {
+		return true, "", nil
+	}
+
+	functionPods, err := lc.getFunctionPods(ctx, function)
+	if err != nil {
+		return false, "", err
+	}
+	// since we are here, it means that we have already checked that the expected number of pods isn't zero
+	// so at least one is expected
+	if functionPods == nil {
+		return false, "", errors.New("No any pods found")
+	}
+
+	// checking that the number of pods is equal to expected replicas, otherwise checking init container
+	// statuses doesn't make sense; need to wait more time
+	if *functionDeployment.Spec.Replicas != int32(len(functionPods.Items)) {
+		notReadyReason = fmt.Sprintf("Not all pod replicas are deployed yet. Expected replicas: %d. Actual replicas: %d",
+			*functionDeployment.Spec.Replicas,
+			len(functionPods.Items))
+		return false, notReadyReason, nil
+	}
+
+	// going through each pod's init containers and check that they all were terminated with exit code 0
+	for _, pod := range functionPods.Items {
+		for _, initContainer := range pod.Status.InitContainerStatuses {
+			switch {
+			case initContainer.State.Terminated != nil:
+				if initContainer.State.Terminated.ExitCode == 0 {
+					continue
+				} else {
+					errorMessage := fmt.Sprintf("Init container has been terminated"+
+						"with non zero error code. ExitCode: %d. Reason %s",
+						initContainer.State.Terminated.ExitCode,
+						initContainer.State.Terminated.Reason,
+					)
+					// if init container is terminated, but exit with non-zero exit code, then we check
+					// pod's restart policy and if it's `Never`, we exit with error; otherwise we wait
+					if pod.Spec.RestartPolicy == v1.RestartPolicyNever {
+						return false, "", errors.New(errorMessage)
+					} else {
+						notReadyReason = fmt.Sprintf("Init container %s was failed with non-zero code, "+
+							"but it will be retried", initContainer.Name)
```

Comment:
```suggestion
						notReadyReason = fmt.Sprintf("Init container %s has failed with non-zero code, "+
							"but it will be restarted", initContainer.Name)
```